access/pagerduty/app.go

```diff
@@ -55,337 +69,537 @@ func (a *App) WaitReady(ctx context.Context) (bool, error) {
 	}
 
-	ctx, log = logger.WithField(ctx, "pd_incident_id", action.IncidentID)
+	// Then, try tro approve the request if user is currently on-call.
+	err := a.tryApproveRequest(ctx, req, data.IncidentID)
+	return trace.NewAggregate(resultErr, trace.Wrap(err))
+}
 
-	pluginData, err := a.getPluginData(ctx, reqID)
-	if err != nil {
-		return trace.Wrap(err)
+func (a *App) onResolvedRequest(ctx context.Context, req types.AccessRequest) error {
+	var notifyErr error
+	if _, err := a.postReviewNotes(ctx, req.GetName(), req.GetReviews()); err != nil {
+		notifyErr = trace.Wrap(err)
+	}
+
+	resolution := Resolution{Reason: req.GetResolveReason()}
+	switch req.GetState() {
+	case types.RequestState_APPROVED:
+		resolution.Tag = ResolvedApproved
+	case types.RequestState_DENIED:
+		resolution.Tag = ResolvedDenied
+	}
+	err := trace.Wrap(a.resolveIncident(ctx, req.GetName(), resolution))
+	return trace.NewAggregate(notifyErr, err)
+}
+
+func (a *App) onDeletedRequest(ctx context.Context, reqID string) error {
+	return a.resolveIncident(ctx, reqID, Resolution{Tag: ResolvedExpired})
+}
+
+func (a *App) getNotifyServiceName(req types.AccessRequest) (string, error) {
+	annotationKey := a.conf.Pagerduty.RequestAnnotations.NotifyService
+	slice, ok := req.GetSystemAnnotations()[annotationKey]
+	if !ok {
+		return "", trace.Errorf("request annotation %q is missing", annotationKey)
+	}
+	var serviceName string
+	if len(slice) > 0 {
+		serviceName = slice[0]
 	}
+	if serviceName == "" {
+		return "", trace.Errorf("request annotation %q is empty", annotationKey)
+	}
+	return serviceName, nil
+}
 
-	if pluginData.PagerdutyData.ID == "" {
-		return trace.Errorf("plugin data is empty")
+func (a *App) tryNotifyService(ctx context.Context, req types.AccessRequest, serviceName string) (PagerdutyData, bool, error) {
+	ctx, _ = logger.WithField(ctx, "pd_service_name", serviceName)
+	service, err := a.pagerduty.FindServiceByName(ctx, serviceName)
+	if err != nil {
+		return PagerdutyData{}, false, trace.Wrap(err)
 	}
 
-	if pluginData.PagerdutyData.ID != action.IncidentID {
-		log.WithField("plugin_data_incident_id", pluginData.PagerdutyData.ID).Debug("plugin_data.incident_id does not match incident.id")
-		return trace.Errorf("incident_id from request's plugin_data does not match")
+	reqID := req.GetName()
+	reqData := RequestData{
+		User:          req.GetUser(),
+		Roles:         req.GetRoles(),
+		Created:       req.GetCreationTime(),
+		RequestReason: req.GetRequestReason(),
 	}
 
-	var userEmail, userName string
-	if userID := action.Agent.ID; userID != "" {
-		agent, err := a.bot.GetUserInfo(ctx, userID)
-		if err != nil {
-			log.WithError(err).Errorf("Cannot get user info by id %q", userID)
-		} else {
-			userEmail = agent.Email
-			userName = agent.Name
+	isNew, err := a.modifyPluginData(ctx, reqID, func(existing *PluginData) (PluginData, bool) {
+		if existing != nil {
+			return PluginData{}, false
 		}
+		return PluginData{RequestData: reqData}, true
+	})
```

Comment:
Can you name the functions passed into `modifyPluginData` here and below? Otherwise it's unclear what these functions are supposed to do.
```suggestion
	createNewIfNotExisting := func(existing *PluginData) (PluginData, bool) {
		if existing != nil {
			return PluginData{}, false
		}
		return PluginData{RequestData: reqData}, true
	}
	isNew, err := a.modifyPluginData(ctx, reqID, isNew)
```

Reply:
I like the elegance of anonymous functions, naming them makes code look more indirect. But for every usage of `modifyPluginData` I added some clarifying comments.